### PR TITLE
Use EventTarget in worker / browser runtimes

### PR DESCRIPTION
**Pull request by anniel-stripe**
## Summary
1. Refactor `_emitter` property on the `Stripe` object to use `EventTarget` instead of `EventEmitter`  in worker environments.
2. Add a base class `PlatformFunctions` that is implemented by `WebPlatformFunctions` and `NodePlatformFunctions`
3. Consolidate checking for streaming data  and buffering logic into a `tryBufferData` platform function, which throws an error in workers if a `ReadableStream` is detected.

## Workarounds because of Node <15
`Event` and `EventTarget` are only available in the global scope [starting with Node 14](https://nodejs.org/api/events.html#class-event:~:text=Changes-,v15.0.0,-The%20Event%20class). Since `_StripeEvent ` needs to extend `Event` (to add the `data` object used by listener functions), I ended up creating a base class for `PlatformFunctions` and a separate `WebPlatformFunctions` that is the only place where `StripeEmitter` is imported.

In the tests I added to `PlatformFunctions.spec.ts`, I also added a check for the Node version so that `WebPlatformFunctions` test do not run in Node <= 14.

## Testing
This gets rid of the last of the build errors when importing stripe-node in Cloudflare Workers without polyfills. I tested this new implementation in Cloudflare Workers / Pages and both emit events as expected.

### CODE REVIEW

File: test/PlatformFunctions.spec.ts [hunk 1]
```diff
@@ -181,5 +184,59 @@ for (const platform in platforms) {
...
+
+        if (!isNodeEnvironment) {
+          before(() => {
+            oldCreateEmitter = platformFunctions.createEmitter;
+
+            // StripeEmitter uses the Event Web API. `createEmitter` needs
+            // to be defined here to avoid reference errors in Node versions <15.
+            platformFunctions.createEmitter = (): StripeEmitter => {
+              return new StripeEmitter(new EventTarget());
+            };
```
Comment (anniel-stripe) [inline test/PlatformFunctions.spec.ts:201]: Is there a better way to share this `createEmitter` implementation with `stripe.worker.ts` without defining it in its own module? Should I just create a new module?

File: src/platform/DefaultPlatformFunctions.ts [hunk 2]
```diff
@@ -38,6 +41,10 @@ class DefaultPlatformFunctions {
     }
     return result === 0;
   }
+
+  createEmitter(): StripeEmitter | EventEmitter {
+    throw new Error('createEmitter not implemented');
```
Comment (pakrym-stripe) [inline src/platform/DefaultPlatformFunctions.ts:46]: Feels like this should return a new instance of StripeEmitter

File: src/stripe.common.ts [hunk 3]
```diff
@@ -203,6 +202,7 @@ Stripe.prototype = {
   on: null!,
   off: null!,
   once: null!,
+  emit: null!,
```
Comment (pakrym-stripe) [inline src/stripe.common.ts:205]: ?

File: src/StripeEmitter.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+class _StripeEvent extends Event {
+  data: any;
+  constructor(eventName: string, data: any) {
+    super(eventName);
+    this.data = data;
+  }
+}
+
+/** Minimal EventEmitter wrapper around EventTarget. */
+class StripeEmitter {
```
Comment (pakrym-stripe) [inline src/StripeEmitter.ts:13]: Let's write some tests that run both on `StripeEmitter` and `EventEmiter`. We need to confirm that `StripeEmitter` and has the same behavior as `EventEmiter`.

File: src/StripeEmitter.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  eventTarget: EventTarget;
+  listenerMapping: Record<string, (...args: any[]) => any>;
+
+  constructor(eventTarget: EventTarget) {
+    this.eventTarget = eventTarget;
+    this.listenerMapping = {};
+  }
+
+  on(eventName: string, listener: (...args: any[]) => void): void {
+    const listenerWrapper = (event: any): void => {
```
Comment (pakrym-stripe) [inline src/StripeEmitter.ts:23]: can we have a better type than any?

File: src/stripe.worker.ts [hunk 6]
```diff
@@ -1,8 +1,16 @@
...
+import StripeEmitter = require('./StripeEmitter');
 
 Stripe.createHttpClient = Stripe.createFetchHttpClient;
 Stripe.webhooks._createCryptoProvider = Stripe.createSubtleCryptoProvider;
 
+// StripeEmitter uses the Event Web API. `Event` is not available
+// in the global scope until Node 15, so we set `createEmitter` here
+// to avoid reference errors.
+Stripe._platformFunctions.createEmitter = (): StripeEmitter => {
+  return new StripeEmitter(new EventTarget());
```
Comment (pakrym-stripe) [inline src/stripe.worker.ts:11]: What fails if we move this logic to DefaultPlatformFunctions?

File: src/utils.ts [hunk 7]
```diff
@@ -252,7 +251,13 @@ const utils = {
    */
   checkForStream: (obj: {file?: {data: unknown}}): boolean => {
     if (obj.file && obj.file.data) {
-      return obj.file.data instanceof EventEmitter;
+      return (
```
Comment (pakrym-stripe) [inline src/utils.ts:254]: Can streams of be represented in non-node environments?

File: src/stripe.worker.ts (hunk 6 above)
Comment (pakrym-stripe) [reply src/stripe.worker.ts:11]: We might need a base type for platform functions and then 2 implementations: one for node and one for workers. In that way the node implementation won't have to inherit all dependencies of the worker implementation.

File: src/utils.ts (hunk 7 above)
Comment (pakrym-stripe) [reply src/utils.ts:254]: Yeah, looks like there is something https://developer.mozilla.org/en-US/docs/Web/API/FileReader#events

**Review by pakrym-stripe — CHANGES_REQUESTED** [review]
1. Let's refactor platform functions a bit so we don't have to push more logic into entry point.
2. Let's add some tests to confirm parity between event emitter implementations.
3. Let's decide what we want to do with file streams in non-node environments.

File: src/platform/DefaultPlatformFunctions.ts (hunk 2 above)
Comment (anniel-stripe) [reply src/platform/DefaultPlatformFunctions.ts]: Updated in latest commit - `createEmitter` returns `StripeEmitter` in `WebPlatformFunctions`

File: src/StripeEmitter.ts (hunk 5 above)
Comment (anniel-stripe) [reply src/StripeEmitter.ts]: Changed to `_StripeEvent` in https://github.com/stripe/stripe-node/pull/1668/commits/fafb866900d20f3001c12537d12305df140e8f20

File: src/stripe.worker.ts (hunk 6 above)
Comment (anniel-stripe) [reply src/stripe.worker.ts]: > What fails if we move this logic to DefaultPlatformFunctions?
This used to fail because it imports `StripeEmitter`, where we extend `Event` for `_StripeEvent`. `Event` is not available globally until Node 15, so this failed for Node 12 and 14 tests.

I followed your suggestion and created a `PlatformFunctions` base type that has `NodePlatformFunctions` and `WebPlatformFunctions` implementations in https://github.com/stripe/stripe-node/pull/1668/commits/fafb866900d20f3001c12537d12305df140e8f20

File: src/StripeEmitter.ts (hunk 4 above)
Comment (anniel-stripe) [reply src/StripeEmitter.ts:15]: This is done in [`PlatformFunctions.spec.ts`](https://github.com/stripe/stripe-node/pull/1668/files#diff-f0d0bd6c151c9a6aff19b6a75d08af3bb6c01c410b4494b9dd1e7144481d0a45L14-L18)

File: src/platform/WebPlatformFunctions.ts [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  }
+
+  /** @override */
+  tryBufferData(data: any): Promise<any> {
+    if (
+      data.file.data.getReader &&
+      typeof data.file.data.getReader === 'function'
+    ) {
+      throw new Error(
+        'Uploading a file as a stream is not supported in non-Node environments. Please open or upvote an issue at github.com/stripe/stripe-node if you use this, detailing your use-case.'
```
Comment (anniel-stripe) [inline src/platform/WebPlatformFunctions.ts:50]: Feel free to suggest a better error message!

File: src/StripeEmitter.ts [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+  constructor(eventName: string, data: any) {
+    super(eventName);
+    this.data = data;
+  }
+}
+
+/** Minimal EventEmitter wrapper around EventTarget. */
+class StripeEmitter {
+  eventTarget: EventTarget;
+  listenerMapping: Record<string, (...args: any[]) => any>;
```
Comment (pakrym-stripe) [inline src/StripeEmitter.ts:17]: nit: let's add an alias for `(...args: any[]) => any` and use it across the file.

Comment (pakrym-stripe) [inline src/StripeEmitter.ts:17]: Instead of being a `Record<string` should it be a `Map<object, ...`? Calling toString on function objects seems a bit hacky.

File: src/Webhooks.ts [hunk 10]
```diff
@@ -60,7 +60,7 @@ type WebhookObject = {
   ) => Promise<WebhookEvent>;
   generateTestHeaderString: (opts: WebhookTestHeaderOptions) => string;
   _createCryptoProvider: () => CryptoProvider | null;
-  _platformFunctions: DefaultPlatformFunctions | null;
+  _platformFunctions: WebPlatformFunctions | null;
```
Comment (pakrym-stripe) [inline src/Webhooks.ts:63]:
```suggestion
  _platformFunctions: PlatformFunctions | null;
```

File: src/platform/PlatformFunctions.ts [hunk 11]
```diff
@@ -1,13 +1,16 @@
...
  * implementations depend on the platform / JS runtime.
  */
-class DefaultPlatformFunctions {
+class PlatformFunctions {
   /**
    * Gets uname with Node's built-in `exec` function, if available.
    */
   getUname(): Promise<string | null> {
-    return Promise.resolve(null);
+    throw new Error('getUname not implemented.');
```
Comment (pakrym-stripe) [inline src/platform/PlatformFunctions.ts:13]: why this change?

File: src/platform/WebPlatformFunctions.ts [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+ * Specializes WebPlatformFunctions using APIs available in Web workers.
+ */
+class WebPlatformFunctions extends PlatformFunctions {
+  /** @override */
+  getUname(): Promise<string | null> {
+    return Promise.resolve(null);
+  }
+
+  /** @override */
+  uuid4(): string {
```
Comment (pakrym-stripe) [inline src/platform/WebPlatformFunctions.ts:14]: do we need a copy here?

File: src/platform/WebPlatformFunctions.ts [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+
+    for (let i = 0; i < len; ++i) {
+      result |= a.charCodeAt(i) ^ b.charCodeAt(i);
+    }
+    return result === 0;
+  }
+
+  /** @override */
+  createEmitter(): StripeEmitter {
+    return new StripeEmitter(new EventTarget());
```
Comment (pakrym-stripe) [inline src/platform/WebPlatformFunctions.ts:40]: Should we create EventTarget inside StripeEmitter ctor?

File: src/stripe.common.ts [hunk 14]
```diff
@@ -53,11 +53,10 @@ Stripe.HttpClient = HttpClient.HttpClient;
 Stripe.HttpClientResponse = HttpClient.HttpClientResponse;
 
 import CryptoProvider = require('./crypto/CryptoProvider');
-import EventEmitter = require('events');
 Stripe.CryptoProvider = CryptoProvider;
 
-import DefaultPlatformFunctions = require('./platform/DefaultPlatformFunctions');
-Stripe._platformFunctions = new DefaultPlatformFunctions();
+import PlatformFunctions = require('./platform/PlatformFunctions');
+Stripe._platformFunctions = new PlatformFunctions();
```
Comment (pakrym-stripe) [inline src/stripe.common.ts:59]: Do we need this? `PlatformFunctions` should never be instantiated. It's a "broken" implementation.

File: test/PlatformFunctions.spec.ts [hunk 15]
```diff
@@ -2,20 +2,35 @@
...
+
+import {expect} from 'chai';
+
+let platforms: Record<string, PlatformFunctions>;
 
-const platforms = {
-  default: new DefaultPlatformFunctions(),
-  node: new NodePlatformFunctions(),
-};
+if (process.versions.node < '15') {
```
Comment (pakrym-stripe) [inline test/PlatformFunctions.spec.ts:14]: Nice

File: test/PlatformFunctions.spec.ts [hunk 16]
```diff
@@ -181,5 +196,158 @@ for (const platform in platforms) {
...
+        // WebPlatformFunctions
+        it('should throw an error in web environments if streaming data is provided', () => {
+          if (process.versions.node < '18') {
+            console.log(
+              `'ReadableStream' is not available in the global scope for ${process.version}, skipping test.`
+            );
+            return;
+          }
+
+          const f = new ReadableStream();
```
Comment (pakrym-stripe) [inline test/PlatformFunctions.spec.ts:293]: This won't work on node as well. Should throw everywhere.

**Review by pakrym-stripe — CHANGES_REQUESTED** [review]
few more nits

File: src/platform/WebPlatformFunctions.ts (hunk 12 above)
Comment (anniel-stripe) [reply src/platform/WebPlatformFunctions.ts:14]: Oh good point, I missed this during the refactor

File: src/StripeEmitter.ts (hunk 9 above)
Comment (anniel-stripe) [reply src/StripeEmitter.ts:17]: This is discouraged by TypeScript, see the following linting error:
```
Don't use `object` as a type. The `object` type is currently hard to use ([see this issue](https://github.com/microsoft/TypeScript/issues/21732)).
Consider using `Record<string, unknown>` instead, as it allows you to more easily inspect and use the keys.eslint[@typescript-eslint/ban-types](https://github.com/typescript-eslint/typescript-eslint/blob/v4.33.0/packages/eslint-plugin/docs/rules/ban-types.md)
```

According to [this SO post](https://stackoverflow.com/a/10858667), it looks like everything gets turned into a string anyways under the hood, so I'd rather be explicit here.

File: test/PlatformFunctions.spec.ts (hunk 16 above)
Comment (anniel-stripe) [reply test/PlatformFunctions.spec.ts:293]: This test doesn't pass in node, but actually I don't think it's worth checking / throwing if the user tries to pass in a ReadableStream. We haven't had any issues / user requests for it, and it would add an additional, unnecessary check in `tryBufferData`.

File: src/StripeEmitter.ts (hunk 9 above)
Comment (pakrym-stripe) [reply src/StripeEmitter.ts:17]: https://stackoverflow.com/questions/10858632/are-functions-valid-keys-for-javascript-object-properties talks about object properties, I was wondering if we should use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map 
>Any value (both objects and [primitive values](https://developer.mozilla.org/en-US/docs/Glossary/Primitive)) may be used as either a key or a value

Comment (pakrym-stripe) [reply src/StripeEmitter.ts:17]:
```
const makeFunc = () => {
    const time = Date.now()
    return () => console.log(time);
}
const a = makeFunc();
const b = makeFunc();

const dict: Record<string, any> = {};
dict[a.toString()] = a;
dict[b.toString()] = b;

Object.values(dict).forEach(f=>f());
```
illustrates the problem with using function.toString()

Comment (anniel-stripe) [reply src/StripeEmitter.ts:17]: Ah sorry, I misunderstood! Thanks for the example, `Map` is definitely what we want. :)

File: test/PlatformFunctions.spec.ts (hunk 16 above)
Comment (pakrym-stripe) [reply test/PlatformFunctions.spec.ts:293]: I'll leave this up to you but we both know what the behavior is when passing a `ReadableStream` on node and it's bad.

File: src/StripeEmitter.ts [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+    this.data = data;
+  }
+}
+
+type Listener = (...args: any[]) => any;
+
+/** Minimal EventEmitter wrapper around EventTarget. */
+class StripeEmitter {
+  eventTarget: EventTarget;
+  listenerMapping: Map<Listener, Listener>;
```
Comment (pakrym-stripe) [inline src/StripeEmitter.ts:19]: the value type can be more precise `(event: _StripeEvent): void`

File: src/platform/NodePlatformFunctions.ts [hunk 18]
```diff
@@ -80,6 +84,39 @@ class NodePlatformFunctions extends DefaultPlatformFunctions {
 
     return super.secureCompare(a, b);
   }
+
+  createEmitter(): EventEmitter {
+    return new EventEmitter();
+  }
+
+  /** @override */
+  tryBufferData(data: StreamingFile): Promise<any> {
```
Comment (pakrym-stripe) [inline src/platform/NodePlatformFunctions.ts:93]: nit: Didn't notice it the first time. `Promise<any>`... Boooo!

File: src/stripe.common.ts [hunk 19]
```diff
@@ -53,11 +53,10 @@ Stripe.HttpClient = HttpClient.HttpClient;
 Stripe.HttpClientResponse = HttpClient.HttpClientResponse;
 
 import CryptoProvider = require('./crypto/CryptoProvider');
-import EventEmitter = require('events');
 Stripe.CryptoProvider = CryptoProvider;
 
-import DefaultPlatformFunctions = require('./platform/DefaultPlatformFunctions');
-Stripe._platformFunctions = new DefaultPlatformFunctions();
+// @ts-ignore
+Stripe._platformFunctions = null;
```
Comment (pakrym-stripe) [inline src/stripe.common.ts:59]: nit: do we need this setter at all?

**Review by pakrym-stripe — APPROVED** [review]

File: src/stripe.common.ts (hunk 19 above)
Comment (anniel-stripe) [reply src/stripe.common.ts:59]: It's just to appease TypeScript 😅 I think functionally it makes more sense to remove this and add a few more `@ts-ignore`'s

Comment (anniel-stripe) [reply src/stripe.common.ts:59]: Do you have a preference?

Comment (pakrym-stripe) [reply src/stripe.common.ts:59]: Let's keep it and try to clean up the entire pattern in a follow up

File: test/PlatformFunctions.spec.ts (hunk 16 above)
Comment (anniel-stripe) [reply test/PlatformFunctions.spec.ts:293]: I agree it's bad since the `ReadableStream` gets stringified, but I think this puts too many guard rails in our library for something users are unlikely to do. Also, `ReadableStream` is not available in Node <18, so we'd have to use a duck-typing check to see if the file data looks like a `ReadableStream`. 🦆

**Review by pakrym-stripe — APPROVED** [review]
